pandas/core/internals/blocks.py

```diff
@@ -454,6 +457,32 @@ def coerce_to_target_dtype(self, other, warn_on_upcast: bool = False) -> Block:
         and will receive the same block
         """
         new_dtype = find_result_type(self.values.dtype, other)
+
+        # In a future version of pandas, the default will be that
+        # setting `nan` into an integer series won't raise.
+        if is_scalar(other) and is_integer_dtype(self.values.dtype):
+            try:
+                is_nan = bool(np.isnan(other))
```

Comment:
I think we should check for all NA values since we hopefully end up with NA only?

Reply:
Setting pd.NA or None also cast to float with NaN nowadays, so agreed that those should be catched here as well. 

But setting pd.NaT currently casts to object dtype, and that one is maybe not something we want to allow? (i.e. continue to warn for that is fine?) So maybe add a `and other is not pd.NaT` or something?

Reply:
yeah I had that earlier, but @phofl brought up that in the future that would also be treated as NA here and so wouldn't upcast

Reply:
Yes, but before you were only checking for nan with `np.isnan`, I think? (plus checking for _not_ being NaT) 
To be clear I think we should also check in general with `isna` so we also check for None and pd.NA. I would maybe just continue disallowing NaT (that currently upcasts to object dtype, not to float)